Tiered storage with NFS blog post (#482)
strimzi/strimzi.github.io: _posts/2025-04-16-tha-various-tiers-of-tiered-storage.md

There are many technologies that might match these requirements.
Apart from the object storage, it can be for example HDFS storage or shared file storage.

NOTE: Apache Kafka uses plugins for the different implementations of the remote storage tier.
But no actual implementations are shipped as part of Apache Kafka itself.
So unless you want to write your own plugin, you have to also make sure that there is already some plugin implemented by someone else and has the required quality.

### Shared storage as a tiered storage

While running high-quality object storage on-premise might be challenging, shared file storage is often already available.
It might have for example the form of Network File System (NFS) volumes.
While the quality of it might differ between the users and their environments, it is in many cases already used by other applications and has the required availability and reliability.

NFS storage matches the requirements we listed in the previous section.
You can use it as a shared read-write-many volume between all your Kafka brokers.

> And it can provide similar benefits to object storage.

Reviewer: Do you think we can also mention the "security/privacy" in NFS, because the data is owned by yourselves, not by the the cloud provider?

Some of the tiered storage benefits can be even amplified when running Kafka on-premise.
For example, when using local persistent volumes as the local storage tier, if you lose the physical server due to hardware failure, you often lose its storage as well.
And tiered storage might make it significantly faster to fully recover your Kafka cluster.

Depending on the size of your Kafka cluster, one of the challenges with NFS storage might be its overall capacity and performance.
But that will be problem mainly for very large Kafka clusters with many brokers and large throughput.

In some cases, shared file storage might also be a viable alternative to consider even when running in public cloud.
Especially if the pricing schema used for the object storage is not a good fit for the way you use Apache Kafka, shared file storage is something you should include in your considerations.

> So how do use NFS as a tiered storage with Strimzi?

Reviewer: how ~~do~~ to use

      # ...
    tieredStorage:
      type: custom
      remoteStorageManager:
        className: io.aiven.kafka.tieredstorage.RemoteStorageManager
        classPath: /opt/kafka/tiered-storage-filesystem/*
        config:
          storage.backend.class: io.aiven.kafka.tieredstorage.storage.filesystem.FileSystemStorage
          storage.root: /mnt/tiered-storage/
          storage.overwrite.enabled: "true"
          chunk.size: "4194304"
  # ...
```

And in all `KafkaNodePool` resources with the `broker` role, we have to mount the NFS volume in `.spec.template` using the additional volumes feature.
The `KafkaNodePool` resources that have only the `controller` role do not need the volume as the would not use tiered storage.

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaNodePool
metadata:
  name: broker
  labels:
    strimzi.io/cluster: my-cluster
spec:
  roles:
    - broker
  # ...
  template:
    pod:
      volumes:
        - name: tiered-storage
          persistentVolumeClaim:
            claimName: tiered-storage-nfs
    kafkaContainer:
      volumeMounts:
        - name: tiered-storage
          mountPath: /mnt/tiered-storage/
```

> Once the Kafka cluster is ready, we can to create a Kafka topic named `tiered-storage-test`.

Reviewer: we can ~~to~~ create a Kafka topic named `tiered-storage-test`

> Keep in mind that in order to use the tiered storage, you always have to enable in the topic as well:

Reviewer: Keep in mind that in order to use the tiered storage, you ~~always~~ have to enable in the topic as well:

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaTopic
metadata:
  name: tiered-storage-test
  labels:
    strimzi.io/cluster: my-cluster
spec:
  partitions: 10
  replicas: 3
  config:
    min.insync.replicas: 2
    retention.bytes: 107374182400 # ~100 Gi
    retention.ms: 604800000 # 7 days
    segment.bytes: 10485760 # ~10MB
    file.delete.delay.ms: 1000
    # Tiered storage configuration
    remote.storage.enable: true
    local.retention.ms: 60000 # 1 minute
    local.retention.bytes: 50000000 # 50 MB
```

To make it easier to see how the tiered storage works, the YAML above tunes some of the configuration options:
* Uses small segment size
* Keeps the local retention very short so that we can better see the log segments being offloaded to the remote storage tier

For a real life use of tiered storage, the values will be likely higher.
But the actual values might differ based on your use-case.

> So make sure to tun them accordingly.

Reviewer: tun -> tune

-rw-r--r--. 1 1000740000 1000740000 10484635 Apr 16 20:24 00000000000002370608.log
-rw-r--r--. 1 1000740000 1000740000      240 Apr 16 20:24 00000000000002370608.snapshot
-rw-r--r--. 1 1000740000 1000740000     1248 Apr 16 20:24 00000000000002370608.timeindex
-rw-r--r--. 1 1000740000 1000740000 10485760 Apr 16 20:24 00000000000002380960.index
-rw-r--r--. 1 1000740000 1000740000  1912190 Apr 16 20:24 00000000000002380960.log
-rw-r--r--. 1 1000740000 1000740000      240 Apr 16 20:24 00000000000002380960.snapshot
-rw-r--r--. 1 1000740000 1000740000 10485756 Apr 16 20:24 00000000000002380960.timeindex
-rw-r--r--. 1 1000740000 1000740000        8 Apr 16 20:16 leader-epoch-checkpoint
-rw-r--r--. 1 1000740000 1000740000       43 Apr 16 20:16 partition.metadata
```

So, where did the old segments go?
They were offloaded to the remote storage tier.
And because we use NFS, we can easily verify that with the `ls` command again.
Inside our NFS volume at the `/mnt/tiered-storage/` path, you will see a subdirectory for each topic that is using tiered storage.
And inside the topic subdirectory another subdirectory for each partition.

> And inside that, you will find all the segments that were offloaded to the remote storage tier:

Reviewer 2: ```suggestion
And inside that, we find all the segments that were offloaded to the remote storage tier:
```

```
kubectl exec -ti my-cluster-broker-0 -- ls -l /mnt/tiered-storage/tiered-storage-test-ZLI5GN1xR1aOqKnwKu5NOQ/0
total 5710364
-rw-r--r--. 1 1000740000 1000740000     6808 Apr 16 20:20 00000000000000000000-eKFr45f2Sca1O0kauyoFBw.indexes
-rw-r--r--. 1 1000740000 1000740000 10484635 Apr 16 20:20 00000000000000000000-eKFr45f2Sca1O0kauyoFBw.log
-rw-r--r--. 1 1000740000 1000740000      736 Apr 16 20:20 00000000000000000000-eKFr45f2Sca1O0kauyoFBw.rsm-manifest
-rw-r--r--. 1 1000740000 1000740000     6172 Apr 16 20:20 00000000000000010352-DmFiyQe8TpOzrpUjPF3rMQ.indexes
-rw-r--r--. 1 1000740000 1000740000 10484650 Apr 16 20:20 00000000000000010352-DmFiyQe8TpOzrpUjPF3rMQ.log
-rw-r--r--. 1 1000740000 1000740000      743 Apr 16 20:20 00000000000000010352-DmFiyQe8TpOzrpUjPF3rMQ.rsm-manifest
-rw-r--r--. 1 1000740000 1000740000     6520 Apr 16 20:20 00000000000000020704-WpjsPsRHTlivhNhxKOpGMA.indexes
-rw-r--r--. 1 1000740000 1000740000 10484680 Apr 16 20:20 00000000000000020704-WpjsPsRHTlivhNhxKOpGMA.log
-rw-r--r--. 1 1000740000 1000740000      744 Apr 16 20:20 00000000000000020704-WpjsPsRHTlivhNhxKOpGMA.rsm-manifest
...
```

> Let's also try to consume the messages to make sure that the broker will correctly retrieve the data from the remote tier and deliver them to the consumer.

Reviewer 2: ```suggestion
We can try to consume the messages to make sure that the broker correctly retrieves the data from the remote tier and delivers them to the consumer.
```

We can use the following command to show us the timestamp of the oldest messages in our topic:

```

> $ kubectl run kafka-consumer -ti --image=quay.io/strimzi/kafka:0.45.0-kafka-3.9.0 --rm=true --restart=Never -- bin/kafka-console-consumer.sh --bootstrap-server my-cluster-kafka-bootstrap:9092 --topic tiered-storage-test --from-beginning --max-messages 10 --property print.timestamp=true --property print.value=false

Reviewer: While using timestamp can implicitly show it's the earliest records we sent, I personally like to use `--property print.offset=true` to show the offset (i.e. offset: 0) to show explicitly it's the offset 0 record. But both ways work for sure. FYI.

Author: I updated it and added the offsets to it as well. Thanks.

CreateTime:1744834829639
CreateTime:1744834829639
CreateTime:1744834829639
CreateTime:1744834829639
CreateTime:1744834829639
CreateTime:1744834829639
CreateTime:1744834829639
CreateTime:1744834829639
CreateTime:1744834829639
CreateTime:1744834829639
Processed a total of 10 messages
pod "kafka-consumer" deleted
```

> When I convert the timestamp to regular time, I can see that it corresponds to `Wed Apr 16 2025 20:20:29`, which is the time when I deployed my producer and it is also the time when the oldest segment in the remote storage tier was created.

Reviewer 2: ```suggestion
Converting the timestamp to regular time, it corresponds to `Wed Apr 16 2025 20:20:29`, which is the time when the producer was deployed and the oldest segment in the remote storage tier was created.
```

> That way, we have confirmed that the data offloaded tot he remote storage tier are provided when the consumer requests them.

Reviewer 2: ```suggestion
This confirms that the data offloaded to the remote storage tier is provided on request to the consumer.
```

### Conclusion

> Hopefully, this blog post demonstrated that object storage is not the only option when it comes to tiered storage in Apache Kafka.

Reviewer 2: ```suggestion
Hopefully, this blog post has demonstrated that object storage is not the only option when it comes to tiered storage in Apache Kafka.
```

Depending on your infrastructure and on your requirements, shared file storage such as NFS is worth considering as well.

> And thanks to that, even if you do not run in public cloud and don't have high-quality object storage available, you can still enjoy the immense benefits of using tiered storage in your Apache Kafka clusters.

Reviewer 2: ```suggestion
If you're running outside the public cloud and don't have high-quality object storage available, you can still enjoy the immense benefits of using tiered storage in your Apache Kafka clusters.
```
